Pytest

djarecka opened the pull request:
I removed `nose` and `unittest` libraries; and rewrote the tests to `pytest`. I still kept `mock` library in a few tests. I tried to not change the tests coverage, they should tests exactly the same functionality. However,` py.test` stops execution after the first `assert` statement that fails. 

This is still not completely finished:
* there are some notes (search for "NOTE_dj") that should be removed, but just kept them for now to think and discuss. 
* there are 2 places that I marked as "FAIL", I had to comment them to not get Fails... not sure if I should change the tests or the tested function.
* a few tests are marked as `xfail`,  they were `skipif(True)` before.
* there are tests that are always skipped (I used the test docker containers)
* `circle.yml` runs `run_pytests.py` instead of `run_nosetests.py`, `run_pytests.py` needs to be improved

I also slightly adjusted docstrings so doctests with `pytest` can be run.

oesteban commented on nipype/algorithms/tests/test_auto_NormalizeProbabilityMapSet.py:
```diff
@@ -11,7 +10,7 @@ def test_NormalizeProbabilityMapSet_inputs():
     for key, metadata in list(input_map.items()):
         for metakey, value in list(metadata.items()):
-            yield assert_equal, getattr(inputs.traits()[key], metakey), value
+            assert getattr(inputs.traits()[key], metakey) == value
```
are these changes done in the tools/checkspecs.py file? If not, they will go away with ```make specs```.

oesteban commented on circle.yml:
```diff
@@ -35,10 +35,10 @@ dependencies:
-        timeout: 2600
-    - docker run -v /etc/localtime:/etc/localtime:ro -e FSL_COURSE_DATA="/root/examples/nipype-fsl_course_data" -v ~/examples:/root/examples:ro -v ~/scratch:/scratch -w /root/src/nipype nipype/nipype_test:py27 /usr/bin/run_nosetests.sh py27 :
-        timeout: 5200
+#    - docker run -v /etc/localtime:/etc/localtime:ro -e FSL_COURSE_DATA="/root/examples/nipype-fsl_course_data" -v ~/examples:/root/examples:ro -v ~/scratch:/scratch -w /root/src/nipype nipype/nipype_test:py35 /usr/bin/run_nosetests.sh py35 :
```
This is actually disabling tests in CircleCI. Could you modify ```run_nosetests.sh``` (https://github.com/nipy/nipype/blob/master/docker/circleci/run_nosetests.sh) so that it calls py.test inside?

If you rename that file, make sure it starts with the prefix ```run_``` so it is automatically added in the docker image.

djarecka replied:
yes, I know :) I was still modifying and testing on other branch without running all examples etc. (so it doesn't run 4h). I've just pushed a new version that runs `run_pytests.sh` (plus all examples; so let's see if it's ok)

djarecka replied:
my `run_pytest.sh` is far from being perfect, I'd need some help (left some notes within the file). Mostly I'm not sure if I should install pytest and other library using this file or docer containers might be changed, so it just works. I also don't know how to recreate the codecov functionality: for now I removed xml files and testing pytest-cov.

satra commented:
see: https://github.com/codecov/example-python

djarecka commented:
@satra : Thanks, will read. In general, I'm using pytest-cov: https://github.com/djarecka/nipype/blob/pytest/docker/circleci/run_pytests.sh#L42
 I'm just not sure about this xml files etc. For now I've removed codecov, but pytest should give a coverage report.

satra commented on docker/circleci/run_pytests.sh:
```diff
@@ -0,0 +1,48 @@
+pip install -U pytest
+pip install pytest-raisesregexp
+pip install pytest-cov
+pip install click
```
`click` is a nipype requirement. and so should `pytest` now.  these should be updated in `nipype/info.py` and `*requirements.txt`

satra commented on nipype/algorithms/tests/test_tsnr.py:
```diff
@@ -1,19 +1,19 @@
-import shutil
 
-class TestTSNR(unittest.TestCase):
+#NOTE_dj: haven't removed mock (have to understand better)
```
can remove this note - `mock` is used to simulate some interaction that may not exist on the system.

satra replied:
however if it's not used in a particular file, it should be removed.

djarecka replied:
I had this note, since I was thinking about trying to use pytest mock and monkeypatching instead of the mock library. I will remove the note and keep the library for now. Can try to change it in different PR, once I understand pytest better.

shoshber replied:
I haven't been super happy with the mock library, so if you do check out pytest mock, tell me what you think!

djarecka replied:
sure, I'll give a second try. Since I've never used any mocking or patching, I had problem with understanding how important the differences between the libraries might be for the project. 
I will open a new PR for it, so it will be easier to review.

shoshber replied:
Cool--it's not a priority or anything :)

satra commented:
@djarecka - this looks great and is almost there.

if you can:
- [ ] merge master and 
- [ ] address any of the comments that require addressing, and 
- [ ] the tests pass, 

we can merge it.

djarecka commented:
@satra:
 after you comments, I've changed:
* updated various requirements files, including `requirements.txt` : removing nosetest, doctest-ignore-unicode and adding pytest; deleting `run_nosetest.`
* changed description of the testing process within `testing_nipype.rst`
* removed installation of pytest from `travis` and `run_pytest.sh` (travis and circle uses `requiremnts.txt`)
* removed my comments that described my changes and some small concerns (might come back to them later)
* removed import nipype.testing when numpy.testing methods are being used 
* changed skipif to use `sys.version_info` and not `TRAVIS_PYTHON_VERSION`, so it works also for `make tests` on a local machine

some issues that I still don't know how to solve (but I believe PR can me merged regardless):
* the only comments I left are related to test failures (`NOTE_dj, FAIL`), two asserts are disabled, since they don't work, either the tests or the code has to be changed.
* there is one doctest that in some situation might not work: https://github.com/djarecka/nipype/blob/pytest/nipype/interfaces/bru2nii.py#L45 
Travis and Circle are ok, but the output might vary.
* there is a file `.coveragerc` that is only used in `Makefile`. Should it be use also in travis and circle?
* I didn't change `install nosetest` to `install pytest` in Dockerfiles, since I'm not sure if this should stay there (some dockerfiles have only `install requirements.txt`)
* don't now what to do with `nipype/fixes/numpy/testing/nosetester.py` and `/nipype/fixes/numpy/testing/noseclasses.py`. can be just removed??
* not sure about `nipype/testing`:  most of the things are not used anymore. I can see only `example_data`, `anatfile` and `funcfile`
* `tools/report_coverage.py` should be changed or removed (had problem running even the original version, so not sure if anybody cares about the file anymore)

satra commented:
we should change the docker files as well.

djarecka commented:
@satra: can remove `pip install nose-cov doctest-ignore-unicode configparser` from Dockerfile_base and Dockerfile_py27. Changing to py.test is not needed IMO, since they run `pip install -r /root/src/nipype/requirements.txt`  (and Dockerfile_py35 doesn't have this part anyway).

Let me know, if you had something else/more in mind.

satra commented:
i think configparser may need to stay for py27

djarecka commented:
@satra all tests were passed after I changed Dockerfiles yesterday, but today I've discovered that my PR can't be merged automatically, since one new test was merged. I rebased again and running tests, but this PR will always have conflict whenever any single test will be changed on the master.

satra commented:
@djarecka - we will merge this as soon as circle passes.

chrisfilo commented:
Hurray! Gratulacje!

On Dec 10, 2016 6:08 AM, "Satrajit Ghosh" <notifications@github.com> wrote:

> Merged #1722 <https://github.com/nipy/nipype/pull/1722>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nipy/nipype/pull/1722#event-889307305>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAOkp5muSi-gdnlPt7GAOh6Sgtk9PrRbks5rGrJ1gaJpZM4K8y3Y>
> .
>

djarecka commented:
@chrisfilo: :-)

djarecka commented on nipype/interfaces/freesurfer/tests/test_preprocess.py:
```diff
@@ -6,11 +6,12 @@
 
-def create_files_in_directory():
+@pytest.fixture()
+def create_files_in_directory(request):
```
it can be, let me do it in a separate PR, I will review some other fixtures. I believe this is not the only one that is often repeated

djarecka commented on nipype/interfaces/fsl/tests/test_BEDPOSTX.py:
```diff
@@ -1,3 +1,6 @@
 from nipype.interfaces.fsl.dti import BEDPOSTX
+
+#NOTE_dj: this test has only import statements
+#NOTE_dj: this is not a AUTO test!
```
I've restored the file, but change the comment. However, I did not notice any new test_auto_BEDPOSTX .

djarecka replied:
checking this issue, I've discovered that I have one auto test that is untracked  - `nipype/algorithms/tests/test_auto_ACompCor.py`. You don't have it?

satra commented:
@djarecka - i can't find the comments you are posting for some reason - but i receive emails. auto-tests are only generated when a corresponding named test is not available. `test_auto_xyz.py` will only be created if `test_xyz.py` does not exist. such custom tests often, but not always, start from an auto test and then modify it.

djarecka commented on nipype/interfaces/fsl/tests/test_dti.py:
```diff
@@ -103,19 +95,20 @@ def setup_tbss():
 
 
-@skipif(skip_dti_tests)
+@pytest.mark.xfail(reason="These tests are skipped until we clean up some of this code") 
```
This test doesn't run properly. Previously had `@skipif(skip_dti_tests)` ,i.e. `@skipif(True)`. Pytest runs tests that are marked as xfail and if they pass pytest would report `xpassed`.

djarecka replied:
the reason was taken from the function`skip_dti_tests`

djarecka commented:
@satra : my point was that after removing `test_XFibres.py` and `test_BEDPOSTX.py` I could not see any differences in auto tests. If you still expect that some new auto tests should be created when these 2 files are missing I can explore it later.

And independent question was about `nipype/algorithms/tests/test_auto_ACompCor.py`. it looks like `checkspecs.py` generates this file for me (it's independent on `test_XFibres.py` and `test_BEDPOSTX.py` files). Was wondering if you don't have this file. It is  not tracked by git.

djarecka commented on nipype/interfaces/fsl/tests/test_maths.py:
```diff
@@ -84,377 +93,330 @@ def test_maths_base(fsl_output_type=None):
 
     # Test the auto naming
     stder = fsl.StdImage(in_file="a.nii")
-    yield assert_equal, stder.cmdline, "fslmaths a.nii -Tstd %s"%os.path.join(testdir, "a_std.nii")
```
ok, I've changed to `stder = fsl.StdImage(in_file="a.nii", output_type='NIFTI')` and test doesn't fail anymore

djarecka commented on nipype/interfaces/tests/test_base.py:
```diff
@@ -545,7 +511,8 @@ class DerivedInterface1(nib.BaseInterface):
     obj = DerivedInterface1()
-    yield assert_not_raises, obj._check_version_requirements, obj.inputs
+    #NOTE_dj: removed yield assert_not_raises
+    obj._check_version_requirements(obj.inputs)
```
yes, if `obj._check_version_requirements(obj.inputs)` raises any exception the test will fail, so  `assert_not_raises` is not needed IMO

djarecka commented:
@satra : the reason why I haven't seen any difference after removing  `test_XFibres.py` and `test_BEDPOSTX.py`, i.e. no new auto tests were created, is here https://github.com/nipy/nipype/blob/master/nipype/interfaces/fsl/dti.py#L428 
`checkspecs.py` consider `XFibres5` and `BEDPOSTX5` as valid classes, and not  `XFibres` and `BEDPOSTX`. 
The auto tests for  `XFibres5` and `BEDPOSTX5` have been created long time ago. They work, but if you want to remove them, I can create `test_BEDPOSTX5.py` and `test_XFibres5.py` and remove the auto tests.
